extensions/warp-ipfs/src/store/conversation.rs

```diff
@@ -34,18 +34,10 @@ pub enum ConversationVersion {
 
 #[derive(Debug, Clone, Serialize, Deserialize, Eq)]
 pub struct ConversationDocument {
-    pub id: Uuid,
+    #[serde(flatten)]
+    pub conversation: Conversation,
```

Comment:
Im not sure about this part. Including `Conversation` in this document would allow reusing the fields from there instead of duplicating them here, but it also makes it hard to separate the internals. Eg, if we wish to use a `Cid` to represent the `recipients` list (unlikely, but could also be possible). then this design would make it impossible as `Conversation` should be agnostic and would be impossible to retrieve that data from `Conversation`. Another example would be if we were to include an image to `Conversation` (eg like a banner, which is may be a feature for groups, and eventually communities), we would have it use a `Cid` internally while externally, it would be in bytes on `Conversation`, so its hard to truly say if this should be here internally or not.

Reply:
> if we wish to use a `Cid` to represent the `recipients` list (unlikely, but could also be possible). then this design would make it impossible as `Conversation` should be agnostic and would be impossible to retrieve that data from `Conversation`.

How would that work with the current design? You still have recipients in the `Conversation` struct. Would recipients in Conversation be just ignored?

> Another example would be if we were to include an image to `Conversation` (eg like a banner, which is may be a feature for groups, and eventually communities), we would have it use a `Cid` internally while externally, it would be in bytes on `Conversation`, so its hard to truly say if this should be here internally or not.

Same question here.

---

Thing is that this change doesn't commit us to eliminate all duplication forever here. If the need arises to duplicate any fields (even the existing ones), then we can still do that.

Reply:
Also I think the representation of concepts should all go in the generic part (`warp`) and not the other way around. We can always make it super generic with traits and associated types, which would allow warp-ipfs to choose any type it wants as long as it satisfies certain trait bounds.

Reply:
> How would that work with the current design? You still have recipients in the Conversation struct. Would recipients in Conversation be just ignored?

When calling `RayGun::get_conversation` (or `RayGun::list_conversations`), internally we can resolve the recipients list if it was pointing to a `Cid` vs just converting from `ConversationDocument` to `Conversation` (as it does now). Under this setup, we have `Conversation` as a field in `ConversationDocument`  and if we were to have `recipients: Cid` (or some field representing pointing to a list of members in a conversation`, that would render the conversation field redundant.

Same if `Conversation` was to have an image supplied to it. On `ConversationDocument`, we may have a field that points to that image that is resolved when calling one of the functions above. While we could mark those fields as default and skip them when serializing, it still will appear to be redundant nonetheless.

Reply:
Thanks for explaining. Could you please also address my previous comment?

Reply:
> Also I think the representation of concepts should all go in the generic part (warp) and not the other way around. We can always make it super generic with traits and associated types, which would allow warp-ipfs to choose any type it wants as long as it satisfies certain trait bounds.

Though would be outside of the scope of this PR (and probably will be done in a far later date for a full overhaul of the api), this would not really solve the overall issue since `ConversationDocument` use `Cid` that points to specific data (which at the time of this comment is used for messages, but would expand out to other items as well). It would not make sense to use `Conversation` in `ConversationDocument` in that case because we would still have fields internally and externally. In my opinion, it would be best to leave `ConversationDocument` as is and map out as needed.